### ISSUE TITLE
Article Layout Margin option for knitr sql code block seems buggy

### ISSUE THREAD

**Issue by eitsupi** [quarto-dev/quarto-cli#4869]
### Bug description

It seems that code blocks as well as the resulting table are placed in the margin.

- OS: Ubuntu 22.04
- Quarto Dev version (latest commit)

````qmd
---
title: test
format:
  html:
    self-contained: true
engine: knitr
---

```{r}
#| echo: false

library(DBI)
library(prqlr)
con <- dbConnect(RSQLite::SQLite(), ":memory:")
dbWriteTable(con, "mtcars", mtcars)
```

```{r, chunk-one}
#| column: margin
mtcars |>
  dplyr::select(cyl) |>
  head(5) |>
  knitr::kable()
```

```{sql}
#| column: margin
#| connection: con

select cyl from mtcars limit 5
```

```{r, chunk-one}
#| column: margin
```
````

![Animation](https://user-images.githubusercontent.com/50911393/225949299-717d1ea0-edbd-4726-8d9f-a4b3f5881cf5.gif)

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [ ] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.
- [X] Please provide the output of `quarto check` so we know which version of quarto and its dependencies you're running.

**Comment by dragonstyle** [comment]
Here's the keep-md - it looks like output from the `sql` cell is evading our wrapping of its output.


````markdown
---
title: test
format:
  html:
    self-contained: true
engine: knitr
keep-md: true
---


::: {.cell}

:::

::: {.cell}

```{.r .cell-code}
mtcars |>
  dplyr::select(cyl) |>
  head(5) |>
  knitr::kable()
```

::: {.cell-output-display}
|                  | cyl|
|:-----------------|---:|
|Mazda RX4         |   6|
|Mazda RX4 Wag     |   6|
|Datsun 710        |   4|
|Hornet 4 Drive    |   6|
|Hornet Sportabout |   8|
:::
:::

::: {.cell}

```{.sql .cell-code}
select cyl from mtcars limit 5
```

<div class="knitsql-table">


Table: 5 records

| cyl|
|---:|
|   6|
|   6|
|   4|
|   6|
|   8|

</div>

:::
````

We're attempting to wrap cell outputs in `::: {.cell-output-display}` but are clearly missing this output for some reason. (The layout system attempts to forward positional information on code cells to the output, but in this case since the output is missing the whole cell is being repositioned).

**Comment by cderv** [comment]
> We're attempting to wrap cell outputs in ::: {.cell-output-display} but are clearly missing this output for some reason

This is because of how the `sql` engine works... The output is a character without specific classes, and this is handled by `sew.character` only. Possibly this is a bug on **knitr** where the output should be a `knit_asis` one but it isn't. 

So we are not wrapping this output currently. 

There is several possible solution that we could do on the quarto side for this engine. Need to think of the best one... 

Obviously fixing **knitr** would solve but we can't patch previous **knitr** version 😓

**Comment by cderv** [comment]
In fact this is not in **knitr** - we need to support asis output a bit broader than today. I'll make a PR.

**Comment by cderv** [comment]
Ok after looking closer into this - this is something that we need to think about. This issue is the hint that we don't support with same behavior 

* A chunk output object of class `knit_asis` - so anything wrapped in `knitr::asis_output()`
* A chunk output object when  `results = "asis"` (or `output = "asis"` in Quarto. )

Take this example 

````markdown
---
title: test
format: html
keep-md: true
engine: knitr
---

```{r}
#| results: asis
#| column: margin
cat("# a header")
```

```{r}
#| column: margin
knitr::asis_output("# a header")
```
````

You'll observe the same behavior as with the `sql` engine output because it is only setting the options `results = "asis"` and not outputing an object of class `knit_asis`. We are wrapping the later but not the former. 

![image](https://user-images.githubusercontent.com/6791940/226016275-63440e7f-26bf-4bd0-b10d-8183c39b17d8.png)


For **knitr** they are the same 

````r
> knitr::knit("test.qmd") |> xfun::file_string()


processing file: test2.qmd
                                                                                                                                                                      
output file: test2.md

---
title: test
format: html
keep-md: true
engine: knitr
---


```r
cat("# a header")
```

# a header


```r
knitr::asis_output("# a header")
```

# a header
````

For Quarto, we treat those two outputs differently. I don't know if this is expected... 🤔 

Do we want to wrap anything that is an output of a knitr chunk ? 
I am not sure we do. 

For this specific `sql` engine issue, we know the default is a table, and that we want to wrap so we can detect the default enclosing div and add our wrapping. but the issue seems more generic. 

Again if we were to wrap the output of `sql` engine with `knitr::asis_output()`, we would not have the issue. But this will require new knitr version and not sure we want that.

**Comment by cscheid** [comment]
Thanks for the full diagnosis!

> Do we want to wrap anything outputing "asis" from a knitr chunk ?

My vote here is "no": we should still reserve `asis` as an escape hatch of some kind, for code that promises to do the right thing. If we wrap anything that's `asis`, then there will be no way for someone to build a completely custom knitr output, since we'll always put _something_ in a `cell-output-display` class. 

That seems like a clear mistake to me.

> For this specific `sql` engine issue, we know the default is a table, and that we want to wrap so we can detect the default enclosing div and add our wrapping. 

I think that's the best practical v1.3 fix, it's just one more fixup. The unfortunate complication here is that knitr appears to output `<div class="...">\n...some markdown...</div>` instead of a markdown fenced div. That means that we can't do it in a Lua filter.

> but the issue seems more generic.

I agree. For v1.4 (or possibly even later), we would like to nail down the protocol for a computational engine outputting markdown. We could then promise that any library that outputs in this format will be treated correctly in all formats for the entirety of the lifetime of the quarto project.

**Comment by cderv** [comment]
> If we wrap anything that's asis, then there will be no way for someone to build a completely custom knitr output, since we'll always put something in a cell-output-display class.

I agree with that, but I think this is what we do currently. Any R function that will wrap its output in `knitr::asis_output()` will see Quarto add the `cell-output-display` class, whereas if it uses `options$results = "asis"`, it will not. It seems like something not wanted... 🤔 

> The unfortunate complication here is that knitr appears to output <div class="...">\n...some markdown...</div> instead of a markdown fenced div. That means that we can't do it in a Lua filter.

Don't we have by default the[ `native_divs` extensions](https://pandoc.org/MANUAL.html#extension-native_divs) ? I believe we can catch this in Lua as a Div

````markdown
---
title: test
format: html
keep-md: true
filters: [ custom.lua ]
engine: knitr
---

```{r}
#| echo: false

library(DBI)
library(prqlr)
con <- dbConnect(RSQLite::SQLite(), ":memory:")
dbWriteTable(con, "mtcars", mtcars)
```

```{sql}
#| connection: con
select cyl from mtcars limit 5
```

````
with `custom.lua`
```lua
Div = function(e)
  if e.classes:includes("knitsql-table") then
    e.content:insert(1, pandoc.Para( pandoc.Strong("Hello from Lua")))
  end
  return e
end
```

So if that is not too late to add it, we can indeed wrap the div `knitsql-table` inside a div of class `cell-output-display`. 

Otherwise if this is too late (regarding the order for filters), this can definitely be done inside R code directly by matching on the div and wrap this content with our output div.

I am thinking doing that in R seems better as this is really R specific, but it seems you think Lua is better - so unsure...

> I agree. For v1.4 (or possibly even later), we would like to nail down the protocol for a computational engine outputting markdown. We could then promise that any library that outputs in this format will be treated correctly in all formats for the entirety of the lifetime of the quarto project.

Agree with this. Probably not 1.4. Should we open an issue to track this idea ?

**Comment by eitsupi** [comment]
Sorry, I posted an obviously inappropriate example...... No `library(prqlr)` needed here! (I updated the example to remove the line)

**Comment by cscheid** [comment]
> I am thinking doing that in R seems better as this is really R specific, but it seems you think Lua is better - so unsure...

You're right, this is better done in R.
